packages/trie/src/checkpointDb.ts

```diff
@@ -0,0 +1,150 @@
+ * which validates inputs and sets encoding type.
+ */
+export class CheckpointDB extends DB {
+  public checkpoints: Checkpoint[]
+
+  /**
+   * Initialize a DB instance. If `leveldb` is not provided, DB
+   * defaults to an [in-memory store](https://github.com/Level/memdown).
+   * @param leveldb - An abstract-leveldown compliant store
+   */
+  constructor(leveldb?: LevelUp) {
+    super(leveldb)
+    // Roots of trie at the moment of checkpoint
+    this.checkpoints = []
+  }
+
+  /**
+   * Is the DB during a checkpoint phase?
+   */
+  get isCheckpoint() {
+    return this.checkpoints.length > 0
+  }
+
+  /**
+   * Adds a new checkpoint to the stack
+   * @param root
+   */
+  checkpoint(root: Buffer) {
+    this.checkpoints.push({ keyValueMap: new Map<string, Buffer>(), root })
+  }
+
+  /**
+   * Commits the latest checkpoint
+   */
+  async commit() {
+    const { keyValueMap } = this.checkpoints.pop()!
+    if (!this.isCheckpoint) {
+      // This was the final checkpoint, we should now commit and flush everything to disk
+      const batchOp: BatchDBOp[] = []
+      keyValueMap.forEach(function (value, key) {
+        if (value === null) {
+          batchOp.push({
+            type: 'del',
+            key: Buffer.from(key, 'hex'),
+          })
+        } else {
+          batchOp.push({
+            type: 'put',
+            key: Buffer.from(key, 'hex'),
+            value,
+          })
+        }
+      })
+      await this.batch(batchOp)
+    } else {
+      // dump everything into the current (higher level) cache
+      const currentKeyValueMap = this.checkpoints[this.checkpoints.length - 1].keyValueMap
+      keyValueMap.forEach(function (value, key) {
+        currentKeyValueMap.set(key, value)
+      })
+    }
+  }
+
+  /**
+   * Reverts the latest checkpoint
+   */
+  async revert() {
+    const { root } = this.checkpoints.pop()!
+    return root
+  }
+
+  /**
+   * Retrieves a raw value from leveldb.
+   * @param key
+   * @returns A Promise that resolves to `Buffer` if a value is found or `null` if no value is found.
+   */
+  async get(key: Buffer): Promise<Buffer | null> {
+    // Lookup the value in our cache. We return the latest checkpointed value (which should be the value on disk)
+    for (let index = this.checkpoints.length - 1; index >= 0; index--) {
+      const value = this.checkpoints[index].keyValueMap.get(key.toString('hex'))
+      if (value !== undefined) {
+        return value
+      }
+    }
+    // Nothing has been found in cache, look up from disk
+    // TODO: put value in cache if we are a checkpoint.
+    return await super.get(key)
+  }
+
+  /**
+   * Writes a value directly to leveldb.
+   * @param key The key as a `Buffer`
+   * @param value The value to be stored
+   */
+  async put(key: Buffer, val: Buffer): Promise<void> {
+    if (this.isCheckpoint) {
+      // put value in cache
+      this.checkpoints[this.checkpoints.length - 1].keyValueMap.set(key.toString('hex'), val)
```

Comment:
There was a weird bug where, if you would `set` the value to just the raw `key.toString()`, then if you'd use `Buffer.from(key)` you'd get a different buffer. So I have settled with hex encoding here, but it should be faster/better if we can just use the raw buffer so we don't have to convert it (maybe has to do with some encoding issue?). Very weird.

Reply:
I think hex encoding is the right option here.

I found in the Buffer [docs](https://nodejs.org/api/buffer.html#buffer_buf_tostring_encoding_start_end) that it says:
> If encoding is 'utf8' and a byte sequence in the input is not valid UTF-8, then each invalid byte is replaced with the replacement character U+FFFD.

so I wonder if that could be related to the bug you encountered.

Another idea I had is that base64 encoding would actually output a shorter string for the key, but I'm unsure if that would make a significant difference in the performance of the map. How large could these checkpoint maps conceivably get?

Reply:
These checkpoint maps can get huge, since we checkpoint the trie before the block and then flush it when the block is done. The only thing I'm pretty scary of is when we hit an extremely bad block which hits a lot of trie items and caches these all, this could probably run into some memory issues... 🤔 

The problem I have with this hex encoding is that we actually convert these raw bytes to strings. So if you'd use one byte (let's say `0xff`) then internally it will be represented as `"ff"` which is 2 bytes, so this doubles the strain on memory. I'll take a look at this encoding, maybe if I explicitly pass `utf8` as encoding it works.